### ISSUE TITLE
GPU support?

### ISSUE THREAD

**Issue by flatstik** [openmopac/mopac#32]
Has anyone contacted  [Prof. Rocha's group](http://www.quantum-chem.pro.br/) and asked the code for the GPU support?

**Comment by godotalgorithm** [comment]
So far, we have not attempted to reactivate the old GPU support. It was not a priority for the initial release, and our development plans include an eventual overhaul of GPU support in MOPAC that is unlikely to reuse any of the old implementation. However, given that development has been slower than expected, it might be worth considering the temporary reactivation of the old GPU support, particularly as a ground-breaking activity in supporting GPUs through a CMake-based build system. The old code should still work with at most very minor and localized changes, but handling all of the GPU-related dependencies through CMake is likely to be a tedious exercise that has to be confronted eventually anyways.

**Comment by flatstik** [comment]
Working with large protein models, I would really appreciate to reactivate the GPU support at least temporarily.

**Comment by godotalgorithm** [comment]
The presently-deactivated GPU support in MOPAC is unlikely to be helpful for large protein models. It only accelerates conventional, cubic-scaling MOPAC calculations [https://doi.org/10.1021/ct3004645], and you become limited by both GPU and CPU memory rather than just CPU memory. This feature can reduce time-to-solution for moderately-sized problems, but it does not ease any memory limitations.

Large protein models should be run with MOZYME rather than conventional MOPAC if at all possible. MOZYME is both faster and more memory efficient than conventional MOPAC for large organic molecules. MOZYME has never had GPU support, it is a purely serial CPU implementation. Rocha's group has published preliminary results on GPU acceleration of MOZYME [https://doi.org/10.1007/s00894-020-04571-6], but that feature has not yet been integrated with or contributed to MOPAC. We have solicited the feature from Prof. Rocha, but development activity in his group is a bit slow right now. Also, I should note that the 1-GPU MOZYME benchmarks presented in Rocha's paper are not actually faster than the 1-core CPU MOZYME solver presently implemented in MOPAC. The "40x" speed-up claimed in the abstract of that paper refers to the 1-core CPU implementation of their new solver algorithm, not the existing MOZYME solver in MOPAC.

**Comment by flatstik** [comment]
So the results [published](http://openmopac.net/manual/gpu.html) by Jimmy were fake to start with?

**Comment by aizvorski** [comment]
@godotalgorithm What type of primitives are needed for GPU MOZYME, is it mostly sparse matrix multiply?  How sparse and how large is it?  

Can this be done with either PyTorch https://pytorch.org/docs/stable/sparse.html or JAX https://jax.readthedocs.io/en/latest/jax.experimental.sparse.html ?

Going down to lower level API, there's cuSPARSE https://docs.nvidia.com/cuda/cusparse/index.html but I think higher level would probably be better.

**Comment by godotalgorithm** [comment]
The existing MOZYME solver is based on an approximate sparsity-preserving matrix factorization process (called "pseudodiagonalization"), and the closest standard sparse matrix method would be ILU factorization. The underlying matrix primitive is Givens rotations with truncation of small matrix elements that are created by the rotations. There are no sparse-matrix/dense-vector operations in MOZYME, which seems to be the primary target of the GPU libraries that you have linked.

This is an unfortunate case of a novel algorithm that isn't well adapted to hardware trends, which is also the case in other parts of MOPAC. There was an optional pseudodiagonalization step in the standard MOPAC solver, which uses fewer operations than full diagonalization. However, it inherently uses level-1 BLAS operations rather than level-3, and so it is much slower in practice these days despite the smaller number of operations and I subsequently deactivated it.

On the topic of GPU support, I will soon be reactivating the old cuBLAS/MAGMA code in MOPAC, contingent on CMake behaving itself. I was originally going to rewrite and modernize all of it, but I have other more urgent development priorities and there have been enough requests now for GPU support that I feel obligated to do something.

**Comment by aizvorski** [comment]
@godotalgorithm Thanks for explaining, that's very interesting.  Givens rotations - basically they're matrix-matrix multiplications with a matrix that just has two off-diagonal elements?  (and then truncation).  

I had a look at cuSPARSE and it only has sparse matrix-dense matrix or vector multiply.  cuSPARSELt https://docs.nvidia.com/cuda/cusparselt/index.html may have sparse-sparse?  

Even if cuSPARSE/cuSPARSELt doesn't have the right operations, I think sparse matrix-sparse matrix multiplication should work fairly efficiently in PyTorch.  It would be more efficient to fuse the multiplication and truncation, but doing it as two separate ops shouldn't be terribly slow.  I could try to benchmark it if you give me a couple of example matrixes for a typical-sized protein.  Other libraries could fuse automatically (possibly pytorch-xla), or would support writing efficient code for a fused multiply-truncate (https://openai.com/blog/triton/) without having to write CUDA code directly.

**Comment by godotalgorithm** [comment]
You'll need to give me some time, but I should be able to provide both some matrix examples of various sizes and also a minimal, self-contained single-core reference implementation of pseudodiagonalization. The main idea is quite simple, but there are some bookkeeping complications with its implementation in MOZYME. What I'll need to extract and translate from MOZYME are the Fock matrices for the initial choice of localized molecular orbitals (LMOs). In effect, these are sparse symmetric matrices partitioned into 4 large blocks, and the solver performs a sequence of Givens rotations to eliminate the 2 off-diagonal matrix blocks, which decouples the remaining diagonal matrix blocks.

**Comment by godotalgorithm** [comment]
@aizvorski I have prepared a [first version](https://github.com/godotalgorithm/sparse-pseudo-diag) of a self-contained proxy for the MOZYME solver. It still needs work, but it is already a reasonable reference for any attempts at GPU acceleration of MOZYME and contains small (Crambin) and mid-size (Chymotrypsin) protein examples.

**Comment by aizvorski** [comment]
@godotalgorithm Thanks Jonathan!  That's amazing work.  I read it through and the code is very readable. I'll need a little bit to understand the sequence of operations in carries out when it runs on the example inputs, and then try to benchmark some of the possible GPU functions.

Meanwhile: it seems there is a sparse matrix - sparse matrix multiply in cuSPARSE after all, as cusparseSpGEMM() https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-function-spgemm   I'm not sure is a sequence of those would be more efficient than a naive approach, eg porting the solver you made to CUDA directly, with each column running on its own CUDA thread.

**Comment by godotalgorithm** [comment]
@aizvorski It is unlikely that using something like cusparseSpGEMM will produce good performance for 2-by-2 rotation matrices. I do expect that hand-written CUDA code would be needed to accelerate the algorithm as-is. However, when I come back to this (maybe in a month or so), I plan to understand why MOZYME is maintaining sparsity more effectively than my solver. One explanation is that it may be using smaller-angle, more incremental rotation matrices, and in that scenario, it may be possible to merge multiple 2-by-2 rotation matrices into a more substantial transformation matrix that could more effectively use these operations.

On the other hand, I would expect that the computation of the transformed Fock matrix (`project_fock`) would translate well to cusparseSpGEMM. My implementation is forming the transformed matrix one column at a time with mat-vec operations, but this is more naturally a sequence of two sparse matrix-matrix products.

**Comment by miroi** [comment]
Hello, any progress in this direction ?

**Comment by godotalgorithm** [comment]
At this point, GPU support is unlikely to return to MOPAC. I no longer have any plans to do it myself on my current career trajectory, so it will depend on contributions from others. There is an engineer at AMD who is currently trying to add multi-threaded CPU support to MOZYME, which is probably the best direction for meaningful performance improvements to MOPAC, but also quite challenging. I don't know of any other active performance-related development for MOPAC.

Meaningful, modern GPU support is a lot of work, and the simple reality is that MOPAC does not have a stable group of developers to perform such work. I am now the only "core" MOPAC developer, and unless some unforeseen opportunities materialize, my professional involvement with MOPAC will end soon. I plan to support MOPAC indefinitely as a hobby, but that will not involve any major feature development.
